quantum/debounce/sym_defer_pr.c

```diff
@@ -21,51 +21,75 @@ DEBOUNCE milliseconds have elapsed since the last change.
 #include "timer.h"
 #include <stdlib.h>
 
+#ifdef PROTOCOL_CHIBIOS
+#    if CH_CFG_USE_MEMCORE == FALSE
+#        error ChibiOS is configured without a memory allocator. Your keyboard may have set `#define CH_CFG_USE_MEMCORE FALSE`, which is incompatible with this debounce algorithm.
+#    endif
+#endif
+
 #ifndef DEBOUNCE
 #    define DEBOUNCE 5
 #endif
 
-static uint16_t last_time;
+// Maximum debounce: 255ms
+#if DEBOUNCE > UINT8_MAX
+#    undef DEBOUNCE
+#    define DEBOUNCE UINT8_MAX
+#endif
+
+typedef uint8_t debounce_counter_t;
+
+#if DEBOUNCE > 0
+static fast_timer_t last_time;
 // [row] milliseconds until key's state is considered debounced.
-static uint8_t* countdowns;
+static debounce_counter_t* debounce_counters;
 // [row]
 static matrix_row_t* last_raw;
 
+#    define DEBOUNCE_ELAPSED 0
+
 void debounce_init(uint8_t num_rows) {
-    countdowns = (uint8_t*)calloc(num_rows, sizeof(uint8_t));
-    last_raw   = (matrix_row_t*)calloc(num_rows, sizeof(matrix_row_t));
+    const size_t debounce_counters_size = num_rows * sizeof(debounce_counter_t);
+    debounce_counters                   = (debounce_counter_t*)malloc(debounce_counters_size);
+    memset(debounce_counters, DEBOUNCE_ELAPSED, debounce_counters_size);
```

Comment:
This `memset` can easily break if this code gets revisited later on. If the type of `debounce_counter_t` gets upgraded at the same time as the value of `DEBOUNCE_ELAPSED` becomes non-zero, both of which are abstracted in this commit and which a developer will reasonably assume can be changed without much fuss, the counters will then get initialized to the wrong value.
```suggestion
    debounce_counters = (debounce_counter_t*)malloc(num_rows * sizeof(debounce_counter_t));
    for (uint8_t r = 0; r <= num_rows; r++) {
        debounce_counters[r] = DEBOUNCE_ELAPSED;
    }
```